### PR TITLE
added implementation of reverse function with its pytests

### PR DESCRIPTION
**Pull request by MohitPatni0731**

### CODE REVIEW

File: LinkedList/test_LinkedList_practice.py [hunk 1]
```diff
@@ -169,4 +169,31 @@ def test_nodes_from_empty_llist():
...
+    linked_list = ll([])
+    linked_list.reverse()
+    assert str(linked_list) == "None"
+
+def test_palindrome1():
+    linked_list = ll([1,2,3,4])
+    assert linked_list.is_palindrome() == False
+
+def test_palindrome2():
+    linked_list = ll([1,1,1,1])
```
Comment (patniharshit) [inline LinkedList/test_LinkedList_practice.py:194]: Good test case.

Missing test case odd length LL:
1 -> 2 -> 3 -> 2 -> 1 -> None

Missing test case LL of length 1:
1 -> None (this is also odd length but still 1 length is a special edge case for this problem)

File: LinkedList/LinkedList_practice.py [hunk 2]
```diff
@@ -228,12 +228,24 @@ def copy(self) -> None:
...
         return copy_llist
 
-    def reverse(self):
+    def reverse(self) -> None:
         """
         Inspired by: https://leetcode.com/problems/remove-linked-list-elements/description/
         
         Reverses the list in place.
+        Time - O(N)
+        Space - O(N)
```
Comment (patniharshit) [inline LinkedList/LinkedList_practice.py:237]: There might be more ways to do reverse, which tradeoff time-space complexities. We have implemented one of the ways. In a interview situation you are supposed to present different ways to solve with their time and space complexity analysis depending on what interview want to optimize.

File: LinkedList/LinkedList_practice.py [hunk 3]
```diff
@@ -251,15 +263,23 @@ def remove_duplicates(self) -> None:
             current.next = next_node
             current = next_node 
 
-    def is_palindrome(self):
+    def is_palindrome(self) -> None:
```
Comment (patniharshit) [inline LinkedList/LinkedList_practice.py:266]: Same as above, there can be more ways of solving this palindrome problem, we have done one way by using reverse.

File: LinkedList/LinkedList_practice.py (hunk 2 above)
Comment (patniharshit) [reply LinkedList/LinkedList_practice.py:237]: There is one way with recursion:

```python
# This method name starts with _ to denote its a private method of this class which is not supposed to be called by code outside of this classs
def _recurse_func(self, node: Node) -> Node:
     # recursion base case - if base case is not present then recursion will keep going infinitely
     if node.next is None:
         self.head = node
         return node

     # this is the recursive part, note that how it calls itself again

     new_node = self._recurse_func(node.next)
     
     # this is the part which changes the next pointer in reverse direction
     new_node.next = node

     return node

def reverse(self):
    ```
    Time complexity: O(N) (each node is traversed just once so for N nodes complexity is linear in O(N))
    Space complexity: O(N) (LL is reversed in place, i.e. it is not copied but due to recursion we take space on computer's stack.
     ```
    last_node = self._recurse_func(self.head)
    
    # Clear the next pointer of last_node so that LL does not form a cycle.
    last_node.next = None
    
    # self.head now points to reversed LL
    return self.head 
```

Implementing recursive functions correctly and determining their time-space complexity is slightly tough so I have implemented this for you to read and understand. We need to learn to implement recursive functions because they form the basis of Dynamic Programming which is very popular topic for CP.

**Review by patniharshit — APPROVED** [review]
One comment about TC, otherwise LGTM.
